src/include/type/expression/abstract_expression.h

```diff
@@ -0,0 +1,191 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+#include "common/hash_util.h"
+#include "common/json.h"
+#include "type/expression/expression_defs.h"
+#include "type/type_id.h"
+
+namespace terrier {
+namespace type {
+namespace expression {
+
+/**
+ * An abstract parser expression. Dumb and immutable.
+ */
+class AbstractExpression {
+ protected:
+  /**
+   * Instantiates a new abstract expression. Because these are logical expressions, everything should be known
+   * at the time of instantiation, i.e. the resulting object is immutable.
+   * @param expression_type what type of expression we have
+   * @param return_value_type the type of the expression's value
+   * @param children the list of children for this node
+   */
+  AbstractExpression(const ExpressionType expression_type, const TypeId return_value_type,
+                     const std::vector<AbstractExpression *> children)
+      : expression_type_(expression_type), return_value_type_(return_value_type) {
+    for (auto &child : children) {
+      children_.emplace_back(std::unique_ptr<AbstractExpression>(child));
+      if (child->is_nullable_) {
```

Comment:
The is_nullable_, has_parameter_ and has_subquery_ fields don't appear to have default initialization anywhere, so they are initialized by the default constructor. The rules around what is initialized for you and what is not, are quite complicated. Therefore, much clearer if we add code to initialize these explicitly.